tests/cli/conftest.py

```diff
@@ -123,3 +126,204 @@ def _run_cli(arguments: List[str]) -> SysCapWithCode:
+
+    def write_ref(self, ref: Path, buf: str) -> bool:
+        if ref.exists():
+            orig = ref.read_text()
+            if orig == buf:
+                return False
+        self.write_file(ref, buf)
+        if self._reporter.verbosity > 0:
+            rel_ref = self.rel(ref)
+            self._reporter.write_line(f"Regenerate {rel_ref}", yellow=True)
+        self._written_refs.append(ref)
+        return True
+
+    def summary(self) -> None:
+        if self._reporter.verbosity == 0:
+            if self._written_refs:
+                self._reporter.write_line("Regenerated files:", yellow=True)
+                for fname in self._written_refs:
+                    rel_ref = self.rel(fname)
+                    self._reporter.write_line(f"  {rel_ref}", yellow=True)
+
+    def diff(self, lft: Guard, rgt: Guard) -> List[str]:
+        # The same as _diff_text from
+        # https://github.com/pytest-dev/pytest/blob/master/src/_pytest/assertion/util.py#L200-L245 plus a few extra lines with additional instructions.  # noqa
+        explanation: List[str] = []
+
+        left = click.unstyle(lft.arg)
+        right = click.unstyle(rgt.arg)
+
+        if self._reporter.verbosity < 1:
+            i = 0  # just in case left or right has zero length
+            for i in range(min(len(left), len(right))):
+                if left[i] != right[i]:
+                    break
+            if i > 42:
+                i -= 10  # Provide some context
+                explanation = [
+                    "Skipping %s identical leading characters in diff, use -v to show"
+                    % i
+                ]
+                left = left[i:]
+                right = right[i:]
+            if len(left) == len(right):
+                for i in range(len(left)):
+                    if left[-i] != right[-i]:
+                        break
+                if i > 42:
+                    i -= 10  # Provide some context
+                    explanation += [
+                        "Skipping {} identical trailing "
+                        "characters in diff, use -v to show".format(i)
+                    ]
+                    left = left[:-i]
+                    right = right[:-i]
+
+        keepends = True
+        if left.isspace() or right.isspace():
+            left = repr(str(left))
+            right = repr(str(right))
+            explanation += [
+                "Strings contain only whitespace, escaping them using repr()"
+            ]
+        # "right" is the expected base against which we compare "left",
+        # see https://github.com/pytest-dev/pytest/issues/3333
+        explanation += [
+            line.strip("\n")
+            for line in ndiff(right.splitlines(keepends), left.splitlines(keepends))
+        ]
+        explanation.append("")
+        explanation.append(f"'cat {self.rel(lft.path)}' to see the test output.")
+        explanation.append(f"'cat {self.rel(rgt.path)}' to see the reference.")
```

Comment:
On failure, pytest displays instructions on how to see the expected and actual texts.
Both are stored in files, the expected version is saved by git.

`pytest --rich-gen` updates all *expected* files (references).
You can do it yourself but automation is very handy.
The changes are pretty visible by PR diffs.